cpp/include/raft/matrix/detail/select_k-inl.cuh

```diff
@@ -116,6 +120,83 @@ inline Algo choose_select_k_algorithm(size_t rows, size_t cols, int k)
   }
 }
 
+template <typename T, typename IdxT>
+void sort_k(
+  raft::resources const& handle, IdxT* indices, T* distances, size_t n_rows, int k, bool select_min)
+{
+  auto stream    = raft::resource::get_cuda_stream(handle);
+  auto out_inds  = raft::make_device_matrix<IdxT, IdxT>(handle, n_rows, k);
+  auto out_dists = raft::make_device_matrix<T, IdxT>(handle, n_rows, k);
+  auto offsets   = raft::make_device_vector<IdxT, IdxT>(handle, n_rows + 1);
+
+  raft::matrix::fill(handle, offsets.view(), (IdxT)k);
+
+  thrust::exclusive_scan(raft::resource::get_thrust_policy(handle),
+                         offsets.data_handle(),
+                         offsets.data_handle() + offsets.size(),
+                         offsets.data_handle(),
+                         0);
+
+  // Determine temporary device storage requirements
+  void* d_temp_storage      = NULL;
+  size_t temp_storage_bytes = 0;
+  if (select_min) {
+    cub::DeviceSegmentedRadixSort::SortPairs(d_temp_storage,
+                                             temp_storage_bytes,
+                                             distances,
+                                             out_dists.data_handle(),
+                                             indices,
+                                             out_inds.data_handle(),
+                                             n_rows * k,
+                                             n_rows,
+                                             offsets.data_handle(),
+                                             offsets.data_handle() + 1);
+  } else {
+    cub::DeviceSegmentedRadixSort::SortPairsDescending(d_temp_storage,
+                                                       temp_storage_bytes,
+                                                       distances,
+                                                       out_dists.data_handle(),
+                                                       indices,
+                                                       out_inds.data_handle(),
+                                                       n_rows * k,
+                                                       n_rows,
+                                                       offsets.data_handle(),
+                                                       offsets.data_handle() + 1);
+  }
+  // Allocate temporary storage
+  cudaMalloc(&d_temp_storage, temp_storage_bytes);
```

Comment:
I believe this will synchronize the device - can we use a `rmm::device_uvector<char>` here for temporary storage (which should also pick up a pool allocator if set) ?

Reply:
> I believe this will synchronize the device - can we use a rmm::device_uvector<char> here for temporary storage (which should also pick up a pool allocator if set) ?

Ah good catch! Yes this wasn't intentional. And I've been preferring the `mdarray` over directly using `rmm` in new code so that we can add a layer between the memory manager and the actual containers (e.g. container policy).